lib/api/v3/attachments/attachments_by_work_package_api.rb

```diff
@@ -0,0 +1,106 @@
+#++
+
+require 'api/v3/attachments/attachment_collection_representer'
+
+module API
+  module V3
+    module Attachments
+      class AttachmentsByWorkPackageAPI < ::API::OpenProjectAPI
+        resources :attachments do
+          helpers do
+            def make_uploaded_file(file, file_name)
+              uploaded_file = Rack::Multipart::UploadedFile.new file[:tempfile].path,
+                                                                file[:type],
+                                                                true
+              # I wish I could set the file name in a better way *sigh*
+              uploaded_file.instance_variable_set(:@original_filename, file_name)
+              uploaded_file
+            end
+
+            def parse_metadata(json)
+              return nil unless json
+
+              # FIXME: we should be using the attachment representer to parse the metadata
+              # We can't because it relies on the underlying :file being magical (e.g. for fileName)
+              begin
+                parsed_json = ::JSON.parse(json)
+              rescue ::JSON::ParserError
+                raise ::API::Errors::InvalidRequestBody.new(I18n.t('api_v3.errors.invalid_json'))
+              end
+
+              file_name = parsed_json['fileName']
+
+              unless file_name
+                raise ::API::Errors::Validation.new(
+                        "fileName #{I18n.t('activerecord.errors.messages.blank')}.")
+              end
+
+              {
+                file_name: file_name,
+                description: (parsed_json['description'] || {})['raw'] || ''
+              }
+            end
+          end
+
+          get do
+            self_path = api_v3_paths.attachments_by_work_package(@work_package.id)
+            attachments = @work_package.attachments
+            AttachmentCollectionRepresenter.new(attachments,
+                                                attachments.count,
+                                                self_path)
+          end
+
+          post do
+            authorize(:edit_work_packages, context: @work_package.project)
+
+            metadata = parse_metadata params[:metadata]
+            file = params[:file]
+
+            unless metadata && file
+              raise ::API::Errors::InvalidRequestBody.new(
+                      I18n.t('api_v3.errors.multipart_body_error'))
+            end
+
+            uploaded_file = make_uploaded_file file, metadata[:file_name]
+            attachment = Attachment.new(file: uploaded_file,
+                                        container: @work_package,
+                                        description: metadata[:description],
+                                        author: current_user)
+
+            unless attachment.save
+              raise ::API::Errors::ErrorBase.create(attachment.errors.dup)
+            end
+
+            ::API::V3::Attachments::AttachmentRepresenter.new(attachment)
+          end
```

Comment:
This class is really cluttered by all the hacks you have too pull of in order to make it work. Any chance we can put those hacks into a separate class so that the controller only has to deal with the request/response flow again?